wallet/store.go

```diff
@@ -1,157 +1,242 @@
+		}
+		if k.key != nil {
+			log.Fatalf("Can only generate key once.")
+		}
+		iterations := k.iterations
+		if iterations == 0 {
+			iterations = Pbkdf2Iterations
+		}
+		k.key = pbkdf2.Key(
+			password,
+			k.salt,
+			iterations,
+			EncKeyLen,
+			Pbkdf2HashFunc,
+		)
+		k.pw = password
 	}
 }
 
 // https://cheatsheetseries.owasp.org/cheatsheets/Secrets_Management_Cheat_Sheet.html#71-encryption-types-to-use
 func (k *WalletKey) encrypt(plaintext []byte) (ciphertext []byte, nonce []byte) {
 	block, err := aes.NewCipher(k.key)
 	cobra.CheckErr(err)
-	// Never use more than 2^32 random nonces with a given key because of the risk of a repeat.
-	// TODO: Remind user to not use the same password to encrypt 4 billion times.
-	nonce = make([]byte, 12)
-	_, err = io.ReadFull(rand.Reader, nonce)
-	cobra.CheckErr(err)
+
+	// Using default options for AES-GCM as recommended by the godoc.
+	// For reference, NonceSize is 12 bytes, and TagSize is 16 bytes:
+	// https://cs.opensource.google/go/go/+/refs/tags/go1.19.2:src/crypto/cipher/gcm.go;l=153-158
 	aesgcm, err := cipher.NewGCM(block)
+	nonce = make([]byte, aesgcm.NonceSize())
+	hash := hmac.New(sha512.New, k.key)
+	hash.Write(plaintext)
+	nonce = hash.Sum(nil)[:aesgcm.NonceSize()]
+
 	cobra.CheckErr(err)
 	ciphertext = aesgcm.Seal(nil, nonce, plaintext, nil)
 	return ciphertext, nonce
 }
-func (k *WalletKey) decrypt(ciphertext []byte, nonce []byte) (plaintext []byte) {
+
+func (k *WalletKey) decrypt(ciphertext []byte, nonce []byte) (plaintext []byte, err error) {
 	block, err := aes.NewCipher(k.key)
 	cobra.CheckErr(err)
 	aesgcm, err := cipher.NewGCM(block)
 	cobra.CheckErr(err)
-	plaintext, err = aesgcm.Open(nil, nonce, ciphertext, nil)
-	cobra.CheckErr(err)
-
-	return plaintext
-}
-
-type WalletOpener interface {
-	Open(path string) (*Wallet, error)
-}
-type WalletExporter interface {
-	Export(path string) error
-}
 
-type ExportableWallet struct {
-	// EncryptedWallet is the encrypted wallet data in base58 encoding.
-	EncryptedWallet string `json:"encrypted_wallet"`
-	// Salt is the salt used to derived the wallet's encryption key in base58 encoding.
-	Salt string `json:"salt"`
-	// Nonce is the nonce used to encrypt the wallet in base58 encoding.
-	Nonce []byte `json:"nonce"`
+	if plaintext, err = aesgcm.Open(nil, nonce, ciphertext, nil); err != nil {
+		return nil, err
+	}
+	return plaintext, nil
 }
 
-type WalletStore struct {
-	wk WalletKey
-}
+func (k *WalletKey) Open(file *os.File) (w *Wallet, err error) {
+	jsonWallet, err := os.ReadFile(file.Name())
+	if err != nil {
+		return nil, err
+	}
+	ew := &EncryptedWalletFile{}
+	if err = json.Unmarshal(jsonWallet, ew); err != nil {
+		return nil, err
+	}
 
-func NewStore(wk *WalletKey) *WalletStore {
-	return &WalletStore{
-		wk: *wk,
+	// set the salt, and warn if it's different
+	var salt [Pbkdf2SaltBytesLen]byte
+	saltTmp, err := hex.DecodeString(ew.Secrets.KDFParams.Salt)
+	copy(salt[:], saltTmp)
+	if err != nil {
+		return nil, err
+	}
```

Comment:
Similar here, avoid the temporary buffer:

```suggestion
	// set the salt, and warn if it's different
	var salt [Pbkdf2SaltBytesLen]byte
	_, err := hex.Decode(salt[:], []byte(ew.Secrets.KDFParams.Salt))
	if err != nil {
		return nil, err
	}
```